lib/committee.js

```diff
@@ -0,0 +1,143 @@
+import createDebug from 'debug'
+
+/** @import {Measurement} from './preprocess.js' */
+/** @import {RetrievalResult, CommitteeCheckError} from './typings.js' */
+
+const debug = createDebug('spark:committee')
+
+const REQUIRED_COMMITTEE_SIZE = 30
+
+export class Committee {
+  /**
+   * @param {Pick<Measurement, 'cid' | 'minerId'>} retrievalTask
+   */
+  constructor ({ cid, minerId }) {
+    this.retrievalTask = { minerId, cid }
+
+    /** @type {Measurement[]} */
+    this.measurements = []
+
+    /** @type {string  | CommitteeCheckError} */
+    this.indexerResult = 'MAJORITY_NOT_FOUND'
+
+    /** @type {RetrievalResult} */
+    this.retrievalResult = 'MAJORITY_NOT_FOUND'
+  }
+
+  /**
+   * @param {number} requiredCommitteeSize
+   * @returns
+   */
+  evaluate (requiredCommitteeSize = REQUIRED_COMMITTEE_SIZE) {
+    debug(
+      'Evaluating task %o with %s accepted measurements',
+      this.retrievalTask,
+      this.measurements.length
+    )
+
+    if (!this.#checkCommitteeSize(requiredCommitteeSize)) return
+    if (!this.#checkMeasuredField('providerId')) return
+    if (!this.#checkMeasuredField('indexerResult')) return
+
+    // TODO
+    // - provider_address,
+    // - protocol,
+
+    if (!this.#checkMeasuredField('retrievalResult')) return
+
+    // TODO
+    // - status_code,
+    // - timeout,
+    // - car size
+    // - car checksum
+    // - car too large
+
+    debug('→ majority agrees on indexerResult=%s retrievalResult=%s', this.indexerResult, this.retrievalResult)
+  }
+
+  /**
+   * @param {number} requiredCommitteeSize
+   * @returns
+   */
+  #checkCommitteeSize (requiredCommitteeSize) {
+    if (this.measurements.length < requiredCommitteeSize) {
+      this.#checkFailed('COMMITTEE_TOO_SMALL')
+      debug('→ committee is too small (size=%s required=%s); retrievalResult=%s',
+        this.measurements.length,
+        requiredCommitteeSize,
+        this.retrievalResult
+      )
+      return false
+    }
+
+    return true
+  }
+
+  /**
+   * @template {keyof Measurement} FIELD_NAME
+   * @param {FIELD_NAME} fieldName
+   */
+  #checkMeasuredField (fieldName) {
+    /** @type {Map<Measurement[FIELD_NAME], number>} */
+    const countsOfValues = new Map()
+    let committeeSize = 0
+    for (const m of this.measurements) {
+      if (m.fraudAssessment !== 'OK') continue
+      committeeSize++
```

Comment:
I think I got this wrong.

Let's consider the following case:

- The committee has 30 measurements
- 20 measurements agree on the same `providerId`
- From those 20 measurements above, 12 agree on the `retrievalResult`
- In the current proposal, we calculate `committeeSize=20` and `majoritySize=12` and conclude that we have a majority we can trust.

That has two problems:
1. We need at least 30 measurements to give us "honest majority". With a committee of 20 measurements, we can no longer be sure about honesty.
2. We will apply several checks; each check can halve the committee size. Towards the end, we can end up with committees containing only a handful of measurements, and we would still treat the majority result as meaningful. That's clearly wrong.

I'll will rework this code to always use the original committee size when determining if we have an absolute majority; and I will also add a test for this.

@patrickwoodhead You know a bit about statistics too, do you agree with my reasoning?

Reply:
This is the first time I'm thinking about individual fields wrt to committees, I thought they were only concerned with "rsr pass" vs "rsr fail". Does it help the design to form committees on every field? (Not saying it doesn't, just asking).

If we're forming a committee on each field, I think this class needs to be refactored so there's one instance per field, as there really is one committee per field.

Reply:
> I think this is indeed problematic, as the result depends on the order the checks have been performed. Each check should always test all measurements instead of checks influencing each other

I think that's not going to work.

The retrieval check has the following steps:
1. Resolve miner ID like `f0...` to provider ID (a libp2p peer id)
2. Query IPNI to find provider records for the CID tested and pick only the records advertised by the provider ID found in the previous step. If no records are found, then abort the algorithm.
3. Use the information in the provider record (multiaddr, protocol) to retrieve the content
4. Calculate checksum of the received CAR file

If nodes resolve `minerId` to different `ProviderID` values, then in the next step - IPNI lookup - they will pick different advertisements and perform the retrieval using different protocols & different hosts to retrieve from.

In your proposal, we will look at the `retrievalResult` field in isolation, ignoring how the retrieval was performed.

The way how I look at majorities: we want to find a majority that reports the same tuple `(providerId, indexerResult, retrievalResult)` - that's what I implemented in this PR. In the future, I want to extend that tuple to include also other fields like `address`, `protocol`, `car_checksum`, etc.

Reply:
What about forming a compound key and using that to find a majority? This way it's a single step process. Eg:

```js
for (const m of measurements) {
  const key = `${m.providerId}${m.indexerResult}${m.retrievalResult}...`
  counts[key]++
}
```

Reply:
We could do that. We will need multiple compound steps, though. One step per the output field we want to derive from the measurements:

- indexerResult requires `(providerId, indexerResult)`
- retrievalResult requires `(providerId, indexerResult, retrievalResult,...)`


There is some duplication between these steps. It's not a problem if we have two steps, but it will become a problem if we add more output fields.

My main concern about this compound-key-based approach is the reduced visibility into the underlying data. 

For example, in my proposal, after we add a check for protocol & provider address, we will be able to tell that the majority agrees on the provider ID and the indexer result, but the committee does not agree on the provider address field.

If we use a compound key, we will know that the majority agrees on the provider ID and the indexer result - thanks to the first step. We won't know why the committee does not agree on the overall outcome.

Another example: car_checksums. In my proposal, we will see that a majority agrees on the provider address and protocol and that the content can be retrieved, but the measurements report different CAR checksums. In your proposal, we would need to analyse the measurements manually to find that information.

Having said that: In https://github.com/filecoin-station/spark-evaluate/pull/304/files#diff-ee367bc5e91f746457be9f758a7f69be7f0bf930e08904cd535cbd3b5e93d5e8, we agreed to use a single coarse-grained code `MAJORITY_NOT_FOUND`. That means we won't see the details I described above in the telemetry and stats. We would get them only when running a dry run with DEBUG logs enabled.

In that light, maybe this reduced visibility is a somewhat lesser problem.

Reply:
> We could do that. We will need multiple compound steps, though. One step per the output field we want to derive from the measurements:
> 
> * indexerResult requires `(providerId, indexerResult)`
> * retrievalResult requires `(providerId, indexerResult, retrievalResult,...)`
> 
> There is some duplication between these steps. It's not a problem if we have two steps, but it will become a problem if we add more output fields.

I think as long as we only need `MAJORITY_NOT_FOUND` this is no problem as one step of checking `(rpoviderId, indexerResult, retrievalResult)` together will cover all cases.

I'm concerned that this multi step process isn't only more complex but also provides results that are harder to trust: Say measurements disagree on two results (indexerResult and retrievalResult), the order in which we execute the checks will tell which result we get. In this case this isn't a problem, because if the indexer result is different, the retrieval result doesn't matter. But are we sure we will always add results that are strictly sequential?

Reply:
Are these stepped per-field committees part of the new Spark spec?

Reply:
> I'm concerned that this multi step process isn't only more complex but also provides results that are harder to trust: Say measurements disagree on two results (indexerResult and retrievalResult), the order in which we execute the checks will tell which result we get. In this case this isn't a problem, because if the indexer result is different, the retrieval result doesn't matter. But are we sure we will always add results that are strictly sequential?

The idea behind this current multi-step is to mirror the steps that Spark checker nodes perform. We verify that after each step of the retrieval testing protocol, we have a majority that agrees on what should have been the outcome of that step.

But I agree with you that this design makes it more difficult to reason about the correctness.

I will try the compound-key alternative and see if I run into any issues.

> Are these stepped per-field committees part of the new Spark spec?

No, I don't think we have any spec for how exactly committees should be formed.

Reply:
Then I believe we need to do that first, before jumping into code, or keep the committee logic as simple as possible

Reply:
I think we need to nail down the conceptual side first before there's more merit in reviewing the code. Otherwise we can decide to change the concept and that causes code rewrites and then that review was mostly meaningless.

Reply:
Marking this thread as resolved, the PR uses the compound-key approach now.